### PR TITLE
engine_rd refactor

### PR DESCRIPTION
**Pull request by mmichal10**
Requires #826

### CODE REVIEW

**Review by rafalste — APPROVED** [review]

File: src/engine/engine_rd.c [hunk 1]
```diff
@@ -28,78 +28,65 @@ static void _ocf_read_generic_hit_complete(struct ocf_request *req, int error)
...
-		if (req->error) {
-			/*
-			 * --- Do not submit this request to write-back-thread.
-			 * Stop it here ---
-			 */
-			req->complete(req, req->error);
+	if (req->error) {
+		/*
+		 * --- Do not submit this request to write-back-thread ---
+		 */
```
Comment (jfckm) [inline src/engine/engine_rd.c:66]: Submit to what? :D

File: src/engine/engine_rd.c [hunk 2]
```diff
@@ -106,21 +106,25 @@ static inline void _ocf_read_generic_submit_miss(struct ocf_request *req)
...
+err_alloc:
 	/* Submit read request to core device. */
 	ocf_submit_volume_req(&req->core->volume, req,
 			_ocf_read_generic_miss_complete);
-
-	return;
-
-err_alloc:
-	_ocf_read_generic_miss_complete(req, -OCF_ERR_NO_MEM);
 }
```
Comment (jfckm) [inline src/engine/engine_rd.c:128]: Something's wrong in here. Previously when alloc/mlock failed we completed with error and submited request if neither failed. Now we always submit request and never complete with error?

**Review by jfckm — CHANGES_REQUESTED** [review]

File: src/engine/engine_rd.c (hunk 2 above)
Comment (mmichal10) [reply src/engine/engine_rd.c:128]: If the buffer allocation fails we can't backfill but we it's not a reason to fail the user's IO

File: src/engine/engine_rd.c (hunk 1 above)
Comment (mmichal10) [reply src/engine/engine_rd.c:66]: Once upon the time... 😆

**Review by jfckm — APPROVED** [review]

**Review by Deixx — COMMENTED** [review]
requires rebase, but otherwise lgtm

**Review by Deixx — APPROVED** [review]

**Review by jfckm — APPROVED** [review]

**Pull request merged**